### ISSUE TITLE
question about Josh's YouTube video "Postgres on Kubernetes" posted on Share Learn channel

### ISSUE THREAD

**Issue by Kayli** [zalando/patroni#2105]
Pretty new to both Postgres and Kubernetes, so have a question about load balancer that you created around 15:00  As a client of such db cluster how can I use a single reliable endpoint to commit write transactions? Was a bit confused why load balancer is readonly.

https://youtu.be/XP_iKW7zB4M?t=894 

Thanks!

**Comment by CyberDem0n** [comment]
Josh is talking about two different services, one read-write, that points to the primary and the second read-only, it load-balance across all replicas.

**Comment by Kayli** [comment]
Thanks for a quick feedback, but this does not explain why load-balance service is readonly. As a client of such db cluster how can I use a single reliable endpoint to commit BOTH read and write transactions?

**Comment by CyberDem0n** [comment]
It seems that you are looking for a silver bullet - the special proxy that based on the query type will send it either to the primary or replicas.
There is no reliable solution to this problem in the middleware. First of all, not every SELECT is a read-only, hence

**Comment by Kayli** [comment]
Appreciate your detailed answer.
Okay, I think that there is clearly some misunderstanding here.
Let's take some example not directly related to databases: when I enter google.com address into a browser, I hit a load balancer that routes my request to the best available node of the google website cluster. This is not dependent on type of the query: it can be readonly query for data or command changing state of the application or both. 
From the perspective of end user accessing such service becomes trivial. 
Conclusion here is that there is definitely a possibility for reliable solution on the level of middleware. 

I do understand that Postgres does not support multimaster out of the box, but this must be somehow solved on a client-library level ... so at least it is well encapsulated and transparent for a client library user.  Am I right?

In other words, can I take some Postgresql client library (e.g. in python) and point it to an endpoint and the rest happens automagically without me caring where is replica and where is master? 
Or you saying that I destined to see all guts and should mess with them on a regular basis?

**Comment by Kayli** [comment]
Hey team, should I open a new ticket to get adequate explanation?

**Comment by CyberDem0n** [comment]
Your example with google is actually very bad, in this case you are connect to the application, not database.

Regarding write-read connection routing - I will just repeat what was already said, only application knows whether the specific query is safe to send to the replica.

There some connection poolers available that are capable to do read-write routing, but they will create more problems than benefits and therefore I will not name them here.

**Comment by Kayli** [comment]
My example is actually fine, as we are connecting to a database application ... but let's not play this game of words, as it's kinda silly ... 'silverbullet' ha, am I vampire or something? 
I'm pretty sure you've got what I meant: I there a client-side library for Postgres that tracks replicas and abstracts away all master-slave jazz or whatever is under the carpet? ;)
I'm looking for a foss Postgres solution for k8 that's fault-tolerant and scalable. Just basic requirements here. I thought that in the video mentioned above such solution will be provided, am I wrong?

**Comment by CyberDem0n** [comment]
> let's not play this game of words, as it's kinda silly ... 'silverbullet' ha, am I vampire or something?

https://en.wikipedia.org/wiki/Silver_bullet
```
The term silver bullet is also a metaphor for a simple, seemingly magical, solution to a difficult problem
```

It exactly describes what you are doing.

> I'm pretty sure you've got what I meant: I there a client-side library for Postgres that tracks replicas and abstracts away all master-slave jazz or whatever is under the carpet? ;)

First of all, there are no slaves in Postgres world.

And now regarding the client library that "abstracts" everything. 
Replicas are not always guaranteed to be in sync with the primary, i.e. results of the same query executed on the primary and on the replica could be different. Typically the replication lag is very small, but it could still be a problem for certain cases.

Let's say you have a financial system, and in the database you keep "accounts", "transactions", and "news". The client transfers money from one account to another, immediately refreshes the page, and sees that nothing has changed, because the query hit the replica, which didn't receive this change yet. Such behavior is a nightmare.
At the same time if someone will see a bit outdated news page - it is absolutely fine.

Now please answer the question, how is the client library supposed to know that reads on the "news" table could be safely sent to the replica, but all accounting operations must always be performed on the primary?

> that's fault-tolerant and scalable. Just basic requirements here. I thought that in the video mentioned above such solution will be provided, am I wrong?

Fault-tolerant? Yes.
Scalable? Also Yes. You can scale vertically by providing more resources (CPU/Memory/VolumeSize) and in addition to that you can scale reads nearly infinitely by adding more replicas to the pool.

**Comment by Kayli** [comment]
Appreciate time you’ve taken to write this. 

> The term silver bullet is also a metaphor for a simple, seemingly magical, solution to a difficult problem
> It exactly describes what you are doing.
> 
There is nothing magical about what I’m describing. If master fails, cluster selects another master. Client library should be able to make that transparent to the end user. Same with slave nodes. 

> Now please answer the question, how is the client library supposed to know that reads on the "news" table could be safely sent to the replica, but all accounting operations must always be performed on the primary?
> 

Oh, it's seems simple: if library makes sure that master and slave nodes endpoints (addresses) are up to date, it gives user a choice. Lets say if ‘client’ is a variable referencing library facade instance, then ‘client.master’ may contain up-to-date master address and ‘client.replicas’ may contain a collection of up-to-date replica addresses. 

Again, as I’m new to Kubernetes, I may not know that k8 may take care of assigning some fixed address to master/slave nodes … so, if it is trivially solved on the level of infrastructure configuration - great, however it wasn’t obvious from the video I’ve watched. But for a case when recreated node gets completely new address, it is reasonable to assume that client library should handle that and make transparent to the end user.

> First of all, there are no slaves in Postgres world.
> 
Yeah, lets invent a new language full of euphemisms and hope that no one notices the obvious ;) 

Are you ppl now scared of M and S words too? So funny ...


> Fault-tolerant? Yes.
> Scalable? Also Yes. You can scale vertically by providing more resources (CPU/Memory/VolumeSize) and in addition to that you can scale reads nearly infinitely by adding more replicas to the pool.
> 

If you don’t encapsulate the process of drifting cluster node addresses, you're just ending up providing half-baked solution. And again, maybe this is somehow trivially solved on k8 level, please educate me on that, will be happy to learn something new.

Thanks
